### PR TITLE
Image pull policy (#1345)

### LINKED ISSUE #1324: Add support for image pull policy
There does not seem to be a way to retrieve the latest image version for a GenericContainer via the library. This is not a problem for the provided application containers, but when using a custom image we are forced to manually run `docker pull` outside of the test project before running the tests to ensure we are using the latest image.

It would be nice if the library could be extended with something like `imagePullPolicy` that would allow users to always pull the latest image version of the GenericContainer by leveraging the `pullImageCmd` API in docker-java.

This would also allow to address this long standing issue in Docker:
https://github.com/moby/moby/issues/13331

And would allow to easily support the behavior that is proposed here:
https://github.com/docker/cli/pull/1498

### PR DESCRIPTION
**Pull request by bsideup**
See #1345 by @dmarkhas

Fixes #1324

### CODE REVIEW

**Comment by bsideup** [comment]
Hey @dmarkhas,

Could you please take a look and tell us if this implementation will cover your use cases? Thanks!

File: core/src/main/java/org/testcontainers/images/AbstractImagePullPolicy.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+public abstract class AbstractImagePullPolicy implements ImagePullPolicy {
+
+    private static final LocalImagesCache LOCAL_IMAGES_CACHE = LocalImagesCache.INSTANCE;
+
+    @Override
+    public boolean shouldPull(DockerImageName imageName) {
+        Logger logger = DockerLoggerFactory.getLogger(imageName.toString());
+
+        // Does our cache already know the image?
+        ImageData imageData = LOCAL_IMAGES_CACHE.get(imageName);
```
Comment (rnorth) [inline core/src/main/java/org/testcontainers/images/AbstractImagePullPolicy.java]: I'm half tempted to suggest having an additional boolean that can be used in line 27, e.g. something like:
```
if (isCached && !shouldPullCached(...
```

Alternatively, perhaps renaming this to `cachedImageData` would be equivalently clear.

File: core/src/main/java/org/testcontainers/images/AbstractImagePullPolicy.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            logger.debug("{} is not in image name cache, updating...", imageName);
+            // Was not in cache, inspecting...
+            imageData = LOCAL_IMAGES_CACHE.refreshCache(imageName).orElse(null);
+        }
+
+        if (imageData != null && !shouldPullCached(imageName, imageData)) {
+            return false;
+        }
+
+        log.trace("Should pull image: {}", imageName);
```
Comment (rnorth) [inline core/src/main/java/org/testcontainers/images/AbstractImagePullPolicy.java]: I have the feeling that we should always log, maybe at debug level, about the pull decision (including if we've decided not to pull!)

**Comment by kostapc** [comment]
Also covers case described in #2048

**Pull request merged**